File: ts_sdk/src/entities/asset-cleanup-report.ts
import { APIEntity, registerEntity } from '../APIEntity';
import { dataContext } from '../FlowSync/context';
import { FSRef } from '../fs/FSRef';
import { DockPointerData } from '../models/DockPointer';
import { mainFileForType } from '../models/asset-editor';

/**
 * AssetCleanupReport entity — a garbage-asset scan summary produced by the
 * asset-cleanup flow (the `asset_cleanup` haiku agent classifying skills and
 * agents under the scan roots). The entity row carries only the headline
 * counts the Home Feed card needs; the full payload (per-finding verdicts +
 * rendered markdown) lives at `asset_ref`
 * (~/.claude/cleanup_reports/<name>/report.json) and the viewer streams it via FSRef.
 */
@registerEntity
export class AssetCleanupReport extends APIEntity<AssetCleanupReport> {
  static type: string = 'asset_cleanup_report';

  generated_at?: string;
  root_count: number = 0;
  finding_count: number = 0;
  garbage_count: number = 0;
  keep_count: number = 0;
  unsure_count: number = 0;
  session_id?: string;
  asset_ref?: string;

  constructor(entity: Partial<AssetCleanupReport> = {}) {
    super(entity);
    this.generated_at = entity.generated_at;
    this.root_count = entity.root_count ?? 0;
    this.finding_count = entity.finding_count ?? 0;
    this.garbage_count = entity.garbage_count ?? 0;
    this.keep_count = entity.keep_count ?? 0;
    this.unsure_count = entity.unsure_count ?? 0;
    this.session_id = entity.session_id;
    this.asset_ref = entity.asset_ref;
  }

  override get dockPointer(): DockPointerData {
    return this.assetEditorPointer() ?? this.defaultDockPointer;
  }

  override get editorDockPointer(): DockPointerData {
    return this.assetEditorPointer() ?? super.editorDockPointer;
  }


  /** FsRef for report.json. Resolves compute node from dataContext. */
  get doc(): FSRef | null {
    const typeId = dataContext.computeNodeTypeId;
    if (!typeId || !this.asset_ref) return null;
    // `asset_ref` is the folder; a row from before the unification may still
    // name the inner file, so append only when it is not already there.
    const base = this.asset_ref.replace(/\\/g, '/').replace(/\/+$/, '');
    const main = mainFileForType(AssetCleanupReport.type, 'report.json') as string;
    return new FSRef(base.endsWith(`/${main}`) ? base : `${base}/${main}`, typeId);
  }
}

File: flow_sdk/builtin/asset_cleanup_report.py
"""AssetCleanupReport entity — a garbage-asset scan summary.

Produced by the asset-cleanup flow (``flow_sdk/asset_cleanup``): the
``asset_cleanup`` haiku agent inventories skills/agents under the scan roots
and classifies each; this entity carries only the headline counts the Home
Feed card needs. The full payload — per-finding verdicts + the rendered
markdown — lives in the entity's ``asset_ref`` file
(``.claude/cleanup_reports/<name>/report.json``), mirroring UsageReport.

``report`` is a create-time ferry (db-excluded): it carries the JSON payload
through save into ``default_body_fn`` (which materializes report.json) and is
never persisted or returned on GET.
"""
from __future__ import annotations

import json
from typing import Optional

from flow_sdk.api.api_types.api_field import APIField, NoDBAPIField
from flow_sdk.core import Entity
from flow_sdk.schema.types import EntityType


class AssetCleanupReport(Entity):
    type: str = APIField(default=EntityType.ASSET_CLEANUP_REPORT.value)
    name: str = APIField("")
    generated_at: Optional[str] = APIField(None)

    # Headline (drives the feed card; full detail is in report.json)
    root_count: int = APIField(0)
    finding_count: int = APIField(0)
    garbage_count: int = APIField(0)
    keep_count: int = APIField(0)
    unsure_count: int = APIField(0)
    session_id: Optional[str] = APIField(None, description="Worker session that produced the scan")

    asset_ref: Optional[str] = APIField(None)
    # Create-time ferry only: JSON text consumed by default_body_fn (which
    # materializes report.json at asset_ref). Never persisted to DB/blob.
    report: Optional[str] = NoDBAPIField(default=None)

    @classmethod
    def from_result(
        cls,
        result,
        *,
        name: str | None = None,
        markdown: str | None = None,
        generated_at: str | None = None,
    ) -> "AssetCleanupReport":
        """Build the entity from an ``AssetCleanupResult`` (single mapping authority)."""
        findings = list(result.findings)
        counts = {verdict: len(group) for verdict, group in result.by_verdict().items()}
        report_name = name or (
            f"cleanup-{generated_at[:10]}" if generated_at else "cleanup-report"
        )
        payload = {
            "roots": list(result.roots),
            "findings": [vars(f) for f in findings],
            "summary": dict(result.summary or counts),
            "markdown": markdown or "",
        }
        return cls(
            name=report_name,
            generated_at=generated_at,
            root_count=len(result.roots),
            finding_count=len(findings),
            garbage_count=counts["garbage"],
            keep_count=counts["keep"],
            unsure_count=counts["unsure"],
            session_id=result.session_id,
            report=json.dumps(payload),
        )
